File: ui/src/types/l4-proxy.ts
// L4 Protocol constants
export const L4_PROTOCOLS = ['tcp', 'udp'] as const;
export type L4Protocol = (typeof L4_PROTOCOLS)[number];

// L4 Matcher type constants
export const L4_MATCHER_TYPES = [
  'any',
  'tls',
  'ssh',
  'postgres',
  'http',
  'rdp',
  'socks5',
  'remote_ip',
  'regexp',
] as const;
export type L4MatcherType = (typeof L4_MATCHER_TYPES)[number];

// L4 Load balancing policy constants
export const L4_LOAD_BALANCING_POLICIES = [
  'round_robin',
  'least_conn',
  'random',
  'first',
  'ip_hash',
] as const;
export type L4LoadBalancingPolicy = (typeof L4_LOAD_BALANCING_POLICIES)[number];

// L4 Proxy protocol version constants
export const L4_PROXY_PROTOCOL_VERSIONS = ['v1', 'v2'] as const;
export type L4ProxyProtocolVersion = (typeof L4_PROXY_PROTOCOL_VERSIONS)[number];

export interface L4Upstream {
  host: string;
  port: number;
  weight?: number;
}

export interface L4Route {
  id: number;
  l4_proxy_id: number;
  priority: number;
  matcher_type: L4MatcherType;
  sni_hostnames?: string[];
  allowed_ip_ranges?: string[];
  regex_pattern?: string;
  upstreams: L4Upstream[];
  load_balancing_policy: L4LoadBalancingPolicy;
  tls_terminate: boolean;
  tls_passthrough: boolean;
  proxy_protocol_version?: L4ProxyProtocolVersion;
  created_at: string;
  updated_at: string;
}

export interface L4Proxy {
  id: number;
  name: string;
  description?: string;
  listen_port: number;
  protocol: L4Protocol;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  routes?: L4Route[];
}

export interface L4ProxyListResponse {
  items: L4Proxy[];
  total: number;
  page: number;
  limit: number;
  total_pages: number;
}

export interface L4ProxyStats {
  total_proxies: number;
  active_proxies: number;
  tcp_proxies: number;
  udp_proxies: number;
  total_routes: number;
  total_upstreams: number;
}

export interface CreateL4UpstreamRequest {
  host: string;
  port: number;
  weight?: number;
}

export interface CreateL4RouteRequest {
  priority: number;
  matcher_type: L4MatcherType;
  sni_hostnames?: string[];
  allowed_ip_ranges?: string[];
  regex_pattern?: string;
  upstreams: CreateL4UpstreamRequest[];
  load_balancing_policy: L4LoadBalancingPolicy;
  tls_terminate: boolean;
  tls_passthrough: boolean;
  proxy_protocol_version?: L4ProxyProtocolVersion;
}

export interface CreateL4ProxyRequest {
  name: string;
  description?: string;
  listen_port: number;
  protocol: L4Protocol;
  is_active: boolean;
  routes?: CreateL4RouteRequest[];
}

export interface UpdateL4ProxyRequest {
  name: string;
  description?: string;
  listen_port: number;
  protocol: L4Protocol;
  is_active: boolean;
  routes?: CreateL4RouteRequest[];
}

File: backend/internal/validation/l4_proxy.go
package validation

// CreateL4ProxyRequestDTO represents the request body for creating an L4 proxy
type CreateL4ProxyRequestDTO struct {
	Name        string                    `json:"name" validate:"required,min=1,max=255"`
	Description *string                   `json:"description,omitempty" validate:"omitempty,max=1000"`
	ListenPort  int                       `json:"listen_port" validate:"required,min=1,max=65535"`
	Protocol    string                    `json:"protocol" validate:"required,oneof=tcp udp"`
	IsActive    *bool                     `json:"is_active,omitempty"` // Defaults to true if not provided
	Routes      []CreateL4RouteRequestDTO `json:"routes,omitempty" validate:"dive"`
}

// UpdateL4ProxyRequestDTO represents the request body for updating an L4 proxy
type UpdateL4ProxyRequestDTO struct {
	Name        string                    `json:"name" validate:"required,min=1,max=255"`
	Description *string                   `json:"description,omitempty" validate:"omitempty,max=1000"`
	ListenPort  int                       `json:"listen_port" validate:"required,min=1,max=65535"`
	Protocol    string                    `json:"protocol" validate:"required,oneof=tcp udp"`
	IsActive    bool                      `json:"is_active"`
	Routes      []CreateL4RouteRequestDTO `json:"routes,omitempty" validate:"dive"`
}

// CreateL4RouteRequestDTO represents a route configuration in L4 proxy requests
type CreateL4RouteRequestDTO struct {
	Priority    int    `json:"priority" validate:"gte=0"`
	MatcherType string `json:"matcher_type" validate:"required,oneof=any tls ssh postgres http rdp socks5 remote_ip regexp"`
	// SNIHostnames are only meaningful for the "tls" matcher (SNI is a TLS
	// concept). They are required when MatcherType == "tls" but are NOT
	// restricted for other matcher types here: setting them on, e.g., the
	// "http" matcher is accepted and then silently ignored at config-build
	// time (the L4 http matcher matches any connection; HTTP host routing is
	// L7). See MapMatcherTypeToL4Matcher in caddy/config/layer4_matchers.go.
	SNIHostnames         []string               `json:"sni_hostnames,omitempty" validate:"omitempty,dive,min=1"`
	AllowedIPRanges      []string               `json:"allowed_ip_ranges,omitempty" validate:"omitempty,dive,cidr"`
	RegexPattern         *string                `json:"regex_pattern,omitempty" validate:"omitempty,min=1,max=500"`
	Upstreams            []L4UpstreamRequestDTO `json:"upstreams" validate:"required,min=1,dive"`
	LoadBalancingPolicy  string                 `json:"load_balancing_policy" validate:"required,oneof=round_robin least_conn random first ip_hash"`
	TLSTerminate         bool                   `json:"tls_terminate"`
	TLSPassthrough       bool                   `json:"tls_passthrough"`
	ProxyProtocolVersion *string                `json:"proxy_protocol_version,omitempty" validate:"omitempty,oneof=v1 v2"`
}

// L4UpstreamRequestDTO represents an upstream server in L4 route requests
type L4UpstreamRequestDTO struct {
	Host   string `json:"host" validate:"required,min=1,max=255"`
	Port   int    `json:"port" validate:"required,min=1,max=65535"`
	Weight *int   `json:"weight,omitempty" validate:"omitempty,min=0,max=100"`
}
